From 05634e8d291b3a911d3785d7fabf3581254e2713 Mon Sep 17 00:00:00 2001
From: Mark Stapp <mjs@cisco.com>
Date: Tue, 22 Jul 2025 16:47:04 -0400
Subject: [PATCH 3/5] zebra: Encode resolved-via field for FPM

Add a number-space for FPM nexthop-group attribute encoding,
and add a value for the resolved-via field. Encode the
resolved-via field for FPM nexthop-group messages.

Signed-off-by: Mark Stapp <mjs@cisco.com>
---
 zebra/kernel_netlink.h | 12 ++++++++++++
 zebra/rt_netlink.c     | 10 ++++++++++
 2 files changed, 22 insertions(+)

diff --git a/zebra/kernel_netlink.h b/zebra/kernel_netlink.h
index d920d25810..5fe0b11758 100644
--- a/zebra/kernel_netlink.h
+++ b/zebra/kernel_netlink.h
@@ -31,6 +31,18 @@ struct zebra_ns;
 #define NL_RCV_PKT_BUF_SIZE     (34 * 1024)
 #define NL_PKT_BUF_SIZE         8192
 
+/* FPM Attributes */
+
+/* FPM Attributes for the nexthop/nexthop-group messages.
+ * We pick a starting value well beyond the current OS/netlink values.
+ */
+enum fpm_nha_attrs {
+	/* Resolved-via nexthop group id; uint32 */
+	NHA_FPM_RESOLVED_VIA = 30,
+
+	NHA_FPM_MAX
+};
+
 extern void netlink_parse_rtattr_flags(struct rtattr **tb, int max,
 				 struct rtattr *rta, int len,
 				 unsigned short flags);
diff --git a/zebra/rt_netlink.c b/zebra/rt_netlink.c
index 4ea53cdb5d..14707516d8 100644
--- a/zebra/rt_netlink.c
+++ b/zebra/rt_netlink.c
@@ -3380,6 +3380,16 @@ ssize_t netlink_nexthop_msg_encode(uint16_t cmd,
 				}
 			}
 
+			/* FPM-specific nexthop-group attributes */
+			if (fpm) {
+				if (nh->resolved_via > 0) {
+					if (!nl_attr_put32(&req->n, buflen,
+							   NHA_FPM_RESOLVED_VIA,
+							   nh->resolved_via))
+						return 0;
+				}
+			}
+
 nexthop_done:
 
 			if (IS_ZEBRA_DEBUG_KERNEL)
-- 
2.50.1 (Apple Git-155)

